### PR TITLE
Add support for multiple gadgets

### PR DESCRIPTION
**Pull request by cjpatton**
This extends the PCP system to support types whose validity circuits have multiple distinct gadgets.

### CODE REVIEW

File: src/pcp.rs [hunk 1]
```diff
@@ -299,114 +304,152 @@ where
...
+        .enumerate()
+        .map(|(idx, g)| ProveShimGadget::new(g, g_calls[idx]))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // Create a buffer for storing the proof. The buffer is longer than the proof itself; the extra
+    // length is to accommodate the computation of each of the proof polynomials.
+    let data_len = (0..shim.len())
+        .map(|idx| shim[idx].arity() + shim[idx].deg() * (1 + g_calls[idx]).next_power_of_two())
+        .reduce(|a, b| a + b)
+        .unwrap();
```
Comment (tgeoghegan) [inline src/pcp.rs:351]:
```suggestion
        .sum();
```
`Iterator::sum()` doesn't return an `Option` so you can remove the unsightly `unwrap()`

File: src/pcp.rs [hunk 2]
```diff
@@ -429,167 +472,219 @@ impl<F: FieldElement> From<Vec<F>> for Proof<F> {
...
+    }
+
+    // Create a buffer for the verifier data. This includes the output of the validity circuit as
+    // well as the intermediate polynomials evaluated at the query randomness `query_rand[idx]`,
+    // and the proof polynomial evaluated at `query_rand[idx]` for the `idx`-th gadget.
+    let data_len = 1
+        + (0..shim.len())
+            .map(|idx| shim[idx].arity() + 1)
+            .reduce(|a, b| a + b)
+            .unwrap();
```
Comment (tgeoghegan) [inline src/pcp.rs:537]:
```suggestion
            .sum();
```
as in `prove()`, to avoid, `unwrap()`

**Review by tgeoghegan — CHANGES_REQUESTED** [review]

File: src/pcp.rs (hunk 1 above)
Comment (cjpatton) [reply src/pcp.rs:351]: Nice!

File: src/pcp.rs (hunk 2 above)
Comment (cjpatton) [reply src/pcp.rs:537]: Done. (This also needed a type declaration, i.e., `.sum::<usize>()`.)

**Review by tgeoghegan — APPROVED** [review]

**Comment by cjpatton** [comment]
Fixed up comments a bit and squashed.

**Pull request merged**